Sources/StreamVideo/WebRTC/WebRTCClient.swift

```diff
@@ -1163,10 +1201,93 @@ class WebRTCClient: NSObject, @unchecked Sendable {
         )
     }
     
+    private func subscribeToInternetConnectionUpdates() {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(handleConnectionStateChange),
+            name: .internetConnectionStatusDidChange,
+            object: nil
+        )
+    }
+    
+    @objc private func handleConnectionStateChange(_ notification: NSNotification) {
+        guard let status = notification.userInfo?[Notification.internetConnectionStatusUserInfoKey] as? InternetConnection.Status else {
+            return
+        }
+        
+        if status == .unavailable {
+            if !isFastReconnecting {
+                isFastReconnecting = true
+            }
+            return
+        }
+        if status.isAvailable && isFastReconnecting, let url = signalChannel?.connectURL {
+            signalChannel = makeWebSocketClient(
+                url: url,
+                apiKey: .init(apiKey),
+                isFastReconnect: true
+            )
+            Task {
+                try await connect(
+                    callSettings: callSettings,
+                    videoOptions: videoOptions,
+                    connectOptions: connectOptions!,
+                    fastReconnect: true
+                )
+                checkFastReconnectionStatus()
+            }
+        }
+    }
+    
+    private func checkFastReconnectionStatus(retries: Int = 0) {
+        DispatchQueue.main.asyncAfter(deadline: .now() + fastReconnectTimeout) { [weak self] in
+            guard let self else { return }
+            if (isPeerConnectionConnecting(publisher, otherNotDisconnected: subscriber)
+                || isPeerConnectionConnecting(subscriber, otherNotDisconnected: publisher))
+                && retries == 0 {
+                log.debug("Still connecting, check again after the interval")
+                checkFastReconnectionStatus(retries: 1)
+                return
+            }
+            self.isFastReconnecting = false
+            let reconnectPublisher = publisher != nil ? publisher?.connectionState != .connected : false
+            let reconnectSubscriber = subscriber != nil ? subscriber?.connectionState != .connected : false
+            let shouldFullyReconnect = reconnectPublisher || reconnectSubscriber
+            if shouldFullyReconnect {
+                log.debug("Fast reconnect failed, doing full reconnect")
+                onSignalConnectionStateChange?(.disconnected(source: .noPongReceived))
+            } else {
+                log.debug("Fast reconnect successfull")
+            }
```

Comment:
```suggestion
self.isFastReconnecting = false

func isPeerConnectionDisconnected(_ peerConnection: PeerConnection?) -> Bool {
    guard let peerConnection else {
        return true
    }

    switch peerConnection.connectionState {
    case .disconnected, .failed:
        return true
    default:
        return false
    }
}

let reconnectPublisher = isPeerConnectionDisconnected(publisher)
let reconnectSubscriber = isPeerConnectionDisconnected(subscriber)

let shouldFullyReconnect = reconnectPublisher || reconnectSubscriber
if shouldFullyReconnect {
    log.debug("Fast reconnect failed, doing full reconnect")
    onSignalConnectionStateChange?(.disconnected(source: .noPongReceived))
} else {
    log.debug("Fast reconnect successfully")
}
```

Android seems to considers that a connection needs to be reconnected if its state is `failed` or `disconnected`. ([reference](https://www.notion.so/stream-wiki/Reconnect-96bb776a1c2140ae9a7306dda4318f26?pvs=4#99eb4d9060144aaeb37ab5d717bd90e4))

Reply:
thanks, this is cleaner, will update it. 👍